src/meta/src/barrier/mod.rs

```diff
@@ -552,27 +552,31 @@ where
             self.in_flight_barrier_nums,
         );
 
-        let mut state = BarrierManagerState::create(self.env.meta_store()).await;
-        if self.enable_recovery {
+        let epoch = if self.enable_recovery {
             // handle init, here we simply trigger a recovery process to achieve the consistency. We
             // may need to avoid this when we have more state persisted in meta store.
 
-            // XXX(bugen): why we need this?
-            let new_epoch = state.in_flight_prev_epoch().next();
+            let latest_snapshot = self.hummock_manager.latest_snapshot();
+            assert_eq!(
+                latest_snapshot.committed_epoch, latest_snapshot.current_epoch,
+                "persisted snapshot must be from a checkpoint barrier"
+            );
+            let prev_epoch = TracedEpoch::new(latest_snapshot.committed_epoch.into());
 
             self.set_status(BarrierManagerStatus::Recovering).await;
-            let span = tracing::info_span!("bootstrap_recovery", new_epoch = new_epoch.value().0);
-            let new_epoch = self.recovery(new_epoch).instrument(span).await;
-            state
-                .update_inflight_prev_epoch(self.env.meta_store(), new_epoch)
-                .await
-                .unwrap();
+            let span = tracing::info_span!("bootstrap_recovery", prev_epoch = prev_epoch.value().0);
+            self.recovery(prev_epoch).instrument(span).await
         } else if self.fragment_manager.has_any_table_fragments().await {
             panic!(
                 "Some streaming jobs already exist in meta, please start with recovery enabled \
             or clean up the metadata using `./risedev clean-data`"
-            );
-        }
+            )
+        } else {
+            // Fresh start.
+            TracedEpoch::new(INVALID_EPOCH.into())
```

Comment:
If this is a fresh start, maybe we can also get the epoch from the latest_snapshot? Though in current code it is also `INVALID_EPOCH` for the first hummock version.

Reply:
Yeah, you're right. The epoch for a fresh start is even not necessarily to be `INVALID_EPOCH`, as long as the table fragments are empty.